File: python/tests/test_io.py
"""python/tests/test_io.py — Tests for processmine_ml.io."""

from __future__ import annotations

import pathlib

import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq
import pytest

from processmine_ml.io import (
    read_eventlog_parquet,
    read_xes,
    validate_eventlog,
    write_eventlog_parquet,
)

FIXTURES = pathlib.Path(__file__).parent.parent.parent / "R" / "tests" / "testthat" / "fixtures"


# ---- helpers ----------------------------------------------------------------

def make_minimal_df() -> pd.DataFrame:
    ts = pd.to_datetime(
        ["2024-01-01 08:00:00", "2024-01-01 09:00:00", "2024-01-02 08:00:00"],
        utc=True,
    ).astype("datetime64[us, UTC]")
    return pd.DataFrame({
        "case_id":   ["C1", "C1", "C2"],
        "activity":  ["A",  "B",  "A"],
        "timestamp": ts,
    })


def make_full_df() -> pd.DataFrame:
    ts = pd.to_datetime(
        ["2024-01-01 08:00:00", "2024-01-01 09:00:00", "2024-01-02 08:00:00"],
        utc=True,
    ).astype("datetime64[us, UTC]")
    st = pd.to_datetime(
        ["2024-01-01 07:50:00", "2024-01-01 08:55:00", "2024-01-02 07:45:00"],
        utc=True,
    ).astype("datetime64[us, UTC]")
    return pd.DataFrame({
        "case_id":         ["C1",       "C1",      "C2"],
        "activity":        ["A",        "B",       "A"],
        "timestamp":       ts,
        "start_timestamp": st,
        "resource":        ["alice",    "bob",     "alice"],
        "lifecycle":       ["complete", "complete","start"],
        "case_attrs":      [{"dept": "sales"}, {"dept": "sales"}, {"dept": "ops"}],
        "event_attrs":     [{"priority": "high"}, {}, {"priority": "low"}],
    })


# ---- validate_eventlog ------------------------------------------------------

def test_validate_eventlog_accepts_valid_minimal():
    df = make_minimal_df()
    result = validate_eventlog(df)
    assert result is df


def test_validate_eventlog_accepts_full_log():
    df = make_full_df()
    result = validate_eventlog(df)
    assert result is df


def test_validate_eventlog_rejects_missing_case_id():
    df = make_minimal_df().drop(columns=["case_id"])
    with pytest.raises(ValueError, match="case_id"):
        validate_eventlog(df)


def test_validate_eventlog_rejects_missing_activity():
    df = make_minimal_df().drop(columns=["activity"])
    with pytest.raises(ValueError, match="activity"):
        validate_eventlog(df)


def test_validate_eventlog_rejects_missing_timestamp():
    df = make_minimal_df().drop(columns=["timestamp"])
    with pytest.raises(ValueError, match="timestamp"):
        validate_eventlog(df)


def test_validate_eventlog_rejects_null_in_required():
    df = make_minimal_df().copy()
    df.loc[0, "case_id"] = None
    with pytest.raises(ValueError, match="[Nn]ull|[Nn][Aa]"):
        validate_eventlog(df)


def test_validate_eventlog_rejects_non_utc_timestamp():
    df = make_minimal_df().copy()
    df["timestamp"] = pd.to_datetime(
        ["2024-01-01 08:00:00", "2024-01-01 09:00:00", "2024-01-02 08:00:00"]
    ).tz_localize("America/New_York")
    with pytest.raises(ValueError, match="[Uu][Tt][Cc]"):
        validate_eventlog(df)


def test_validate_eventlog_rejects_timestamp_before_start():
    df = make_minimal_df().copy()
    # start_timestamp > timestamp
    df["start_timestamp"] = df["timestamp"] + pd.Timedelta(hours=1)
    with pytest.raises(ValueError, match="start_timestamp"):
        validate_eventlog(df)


def test_validate_eventlog_rejects_bad_lifecycle():
    df = make_minimal_df().copy()
    df["lifecycle"] = ["complete", "complete", "INVALID"]
    with pytest.raises(ValueError, match="lifecycle"):
        validate_eventlog(df)


# ---- read_xes ---------------------------------------------------------------

def test_read_xes_parses_fixture():
    path = FIXTURES / "tiny.xes"
    log = read_xes(path)

    assert isinstance(log, pd.DataFrame)
    assert len(log) == 5
    assert {"case_id", "activity", "timestamp"}.issubset(log.columns)
    assert str(log["timestamp"].dtype) == "datetime64[us, UTC]"


def test_read_xes_maps_keys_correctly():
    path = FIXTURES / "tiny.xes"
    log = read_xes(path)

    assert sorted(log["case_id"].unique()) == ["case-1", "case-2", "case-3"]
    assert sorted(log["activity"].unique()) == ["approve", "reject", "submit"]
    alice_row = log[(log["case_id"] == "case-1") & (log["activity"] == "submit")]
    assert alice_row["resource"].iloc[0] == "alice"


def test_read_xes_captures_case_attrs():
    path = FIXTURES / "tiny.xes"
    log = read_xes(path)

    assert "case_attrs" in log.columns
    case1 = log[log["case_id"] == "case-1"]["case_attrs"].iloc[0]
    assert case1.get("department") == "sales"


# ---- write/read round-trip --------------------------------------------------

def test_roundtrip_required_columns(tmp_path):
    df = make_minimal_df()
    out = tmp_path / "log.parquet"

    write_eventlog_parquet(df, out)
    back = read_eventlog_parquet(out)

    assert list(back["case_id"]) == list(df["case_id"])
    assert list(back["activity"]) == list(df["activity"])
    pd.testing.assert_series_equal(back["timestamp"], df["timestamp"], check_names=False)


def test_roundtrip_optional_columns(tmp_path):
    df = make_full_df()
    out = tmp_path / "log.parquet"

    write_eventlog_parquet(df, out)
    back = read_eventlog_parquet(out)

    assert list(back["resource"]) == list(df["resource"])
    assert list(back["lifecycle"]) == list(df["lifecycle"])
    pd.testing.assert_series_equal(
        back["start_timestamp"], df["start_timestamp"], check_names=False
    )
    # attrs round-trip as dicts
    assert back["case_attrs"].iloc[0] == {"dept": "sales"}
    assert back["event_attrs"].iloc[0] == {"priority": "high"}
    assert back["event_attrs"].iloc[1] == {}


def test_parquet_carries_schema_version(tmp_path):
    df = make_minimal_df()
    out = tmp_path / "log.parquet"

    write_eventlog_parquet(df, out)
    meta = pq.read_schema(out).metadata
    assert meta[b"schema_version"] == b"1.0"


def test_read_parquet_rejects_unknown_schema_version(tmp_path):
    df = make_minimal_df()
    ts = pa.array(df["timestamp"].values, type=pa.timestamp("us", tz="UTC"))
    table = pa.table({
        "case_id":   pa.array(df["case_id"]),
        "activity":  pa.array(df["activity"]),
        "timestamp": ts,
    })
    table = table.replace_schema_metadata({"schema_version": "2.0"})
    pq.write_table(table, tmp_path / "bad.parquet")

    with pytest.raises(ValueError, match="[Ss]chema.version|version"):
        read_eventlog_parquet(tmp_path / "bad.parquet")

File: R/tests/testthat/test-io.R
# ---- helpers ----------------------------------------------------------------

make_minimal_log <- function() {
  tibble::tibble(
    case_id   = c("C1", "C1", "C2"),
    activity  = c("A", "B", "A"),
    timestamp = as.POSIXct(
      c("2024-01-01 08:00:00", "2024-01-01 09:00:00", "2024-01-02 08:00:00"),
      tz = "UTC"
    )
  )
}

make_full_log <- function() {
  tibble::tibble(
    case_id         = c("C1", "C1", "C2"),
    activity        = c("A", "B", "A"),
    timestamp       = as.POSIXct(
      c("2024-01-01 08:00:00", "2024-01-01 09:00:00", "2024-01-02 08:00:00"),
      tz = "UTC"
    ),
    start_timestamp = as.POSIXct(
      c("2024-01-01 07:50:00", "2024-01-01 08:55:00", "2024-01-02 07:45:00"),
      tz = "UTC"
    ),
    resource  = c("alice", "bob", "alice"),
    lifecycle = c("complete", "complete", "start"),
    case_attrs  = list(
      list(dept = "sales"),
      list(dept = "sales"),
      list(dept = "ops")
    ),
    event_attrs = list(
      list(priority = "high"),
      list(),
      list(priority = "low")
    )
  )
}

# ---- validate_eventlog ------------------------------------------------------

test_that("validate_eventlog accepts a valid minimal log", {
  log <- make_minimal_log()
  expect_invisible(validate_eventlog(log))
})

test_that("validate_eventlog accepts a full log with optional columns", {
  log <- make_full_log()
  expect_invisible(validate_eventlog(log))
})

test_that("validate_eventlog rejects missing case_id column", {
  log <- make_minimal_log()
  log$case_id <- NULL
  expect_error(validate_eventlog(log), "case_id")
})

test_that("validate_eventlog rejects missing activity column", {
  log <- make_minimal_log()
  log$activity <- NULL
  expect_error(validate_eventlog(log), "activity")
})

test_that("validate_eventlog rejects missing timestamp column", {
  log <- make_minimal_log()
  log$timestamp <- NULL
  expect_error(validate_eventlog(log), "timestamp")
})

test_that("validate_eventlog rejects NA in required columns", {
  log <- make_minimal_log()
  log$case_id[1] <- NA_character_
  expect_error(validate_eventlog(log), "NA")
})

test_that("validate_eventlog rejects non-UTC timestamp", {
  log <- make_minimal_log()
  log$timestamp <- as.POSIXct(
    c("2024-01-01 08:00:00", "2024-01-01 09:00:00", "2024-01-02 08:00:00"),
    tz = "America/New_York"
  )
  expect_error(validate_eventlog(log), "[Uu][Tt][Cc]")
})

test_that("validate_eventlog rejects timestamp before start_timestamp", {
  log <- make_minimal_log()
  log$start_timestamp <- log$timestamp + 3600  # start > end
  expect_error(validate_eventlog(log), "start_timestamp")
})

test_that("validate_eventlog rejects invalid lifecycle value", {
  log <- make_minimal_log()
  log$lifecycle <- c("complete", "complete", "INVALID")
  expect_error(validate_eventlog(log), "lifecycle")
})

# ---- read_xes ---------------------------------------------------------------

test_that("read_xes parses the tiny fixture", {
  path <- testthat::test_path("fixtures", "tiny.xes")
  log  <- read_xes(path)

  expect_s3_class(log, "tbl_df")
  expect_equal(nrow(log), 5L)
  expect_true(all(c("case_id", "activity", "timestamp") %in% names(log)))
  expect_equal(attr(log$timestamp, "tzone"), "UTC")
})

test_that("read_xes maps XES keys correctly", {
  path <- testthat::test_path("fixtures", "tiny.xes")
  log  <- read_xes(path)

  expect_equal(sort(unique(log$case_id)), c("case-1", "case-2", "case-3"))
  expect_equal(sort(unique(log$activity)), c("approve", "reject", "submit"))
  expect_true("resource" %in% names(log))
  expect_equal(log$resource[log$case_id == "case-1" & log$activity == "submit"], "alice")
})

test_that("read_xes captures non-standard attributes in case_attrs/event_attrs", {
  path <- testthat::test_path("fixtures", "tiny.xes")
  log  <- read_xes(path)

  expect_true("case_attrs" %in% names(log))
  # case-1 has department = sales
  ca <- log$case_attrs[log$case_id == "case-1"][[1]]
  expect_equal(ca[["department"]], "sales")
})

# ---- write/read_eventlog_parquet --------------------------------------------

test_that("write_eventlog_parquet + read_eventlog_parquet round-trips required columns", {
  log  <- make_minimal_log()
  tmp  <- tempfile(fileext = ".parquet")
  on.exit(unlink(tmp))

  write_eventlog_parquet(log, tmp)
  back <- read_eventlog_parquet(tmp)

  expect_equal(back$case_id,   log$case_id)
  expect_equal(back$activity,  log$activity)
  expect_equal(back$timestamp, log$timestamp)
})

test_that("write_eventlog_parquet + read_eventlog_parquet round-trips all optional columns", {
  log  <- make_full_log()
  tmp  <- tempfile(fileext = ".parquet")
  on.exit(unlink(tmp))

  write_eventlog_parquet(log, tmp)
  back <- read_eventlog_parquet(tmp)

  expect_equal(back$resource,        log$resource)
  expect_equal(back$lifecycle,       log$lifecycle)
  expect_equal(back$start_timestamp, log$start_timestamp)
  # attrs columns round-trip as named character vectors
  expect_equal(back$case_attrs[[1]], c(dept = "sales"))
  expect_equal(back$event_attrs[[1]], c(priority = "high"))
  expect_equal(length(back$event_attrs[[2]]), 0L)
})

test_that("Parquet file carries schema_version metadata", {
  log <- make_minimal_log()
  tmp <- tempfile(fileext = ".parquet")
  on.exit(unlink(tmp))

  write_eventlog_parquet(log, tmp)
  tbl  <- arrow::read_parquet(tmp, as_data_frame = FALSE)
  meta <- tbl$schema$metadata
  expect_equal(meta[["schema_version"]], "1.0")
})

test_that("read_eventlog_parquet rejects unknown major schema version", {
  log <- make_minimal_log()
  tmp <- tempfile(fileext = ".parquet")
  on.exit(unlink(tmp))

  # Write valid parquet, then overwrite schema_version metadata to 2.0
  write_eventlog_parquet(log, tmp)
  tbl <- arrow::read_parquet(tmp, as_data_frame = FALSE)
  tbl <- tbl$ReplaceSchemaMetadata(list(schema_version = "2.0"))
  arrow::write_parquet(tbl, tmp)

  expect_error(read_eventlog_parquet(tmp), "[Ss]chema.version|version")
})
